File: EcosystemSimulation/src/shmp/simulation/space/resource/dependency/AvoidTiles.java
package shmp.simulation.space.resource.dependency;

import shmp.simulation.space.tile.Tile;
import shmp.simulation.space.resource.Resource;

import java.util.Set;


public class AvoidTiles implements ResourceDependency {
    private Set<Tile.Type> badTypes;

    public AvoidTiles(Set<Tile.Type> badTypes) {
        this.badTypes = badTypes;
    }

    @Override
    public double satisfactionPercent(Tile tile, Resource resource) {
        return hasNeeded(tile) ? 1 : 0;
    }

    @Override
    public boolean isNecessary() {
        return true;
    }

    @Override
    public boolean isPositive() {
        return true;
    }

    @Override
    public boolean isResourceNeeded() {
        return false;
    }

    @Override
    public boolean hasNeeded(Tile tile) {
        return !badTypes.contains(tile.getType());
    }
}

File: EcosystemSimulation/src/io/tashtabash/simulation/space/resource/dependency/AvoidTiles.kt
package io.tashtabash.simulation.space.resource.dependency

import io.tashtabash.simulation.space.resource.Resource
import io.tashtabash.simulation.space.tile.Tile


class AvoidTiles(var badTypes: Set<Tile.Type?>) : ResourceDependency {
    override fun satisfactionPercent(tile: Tile, resource: Resource, isSafe: Boolean): Double =
            if (hasNeeded(tile)) 1.0
            else 0.0

    override val isNecessary = true

    override val isPositive = true

    override val isResourceNeeded = false

    override fun hasNeeded(tile: Tile) = !badTypes.contains(tile.type)
}
